java/client/src/main/java/com/vesoft/nebula/client/graph/meta/MetaClient.java

```diff
@@ -0,0 +1,561 @@
+        ListSpacesResp response;
+        try {
+            response = client.listSpaces(request);
+        } catch (TException e) {
+            LOGGER.error(String.format("List Spaces Error: %s", e.getMessage()));
+            return Lists.newLinkedList();
+        }
+        if (response.getCode() == ErrorCode.SUCCEEDED) {
+            return response.getSpaces();
+        } else {
+            LOGGER.error(String.format("List Spaces Error Code: %d", response.getCode()));
+            return Lists.newLinkedList();
+        }
+    }
+
+
+    /**
+     * get all parts and part's location
+     *
+     * @param spaceName nebula graph space
+     * @return empty map if exception happen
+     */
+    private Map<Integer, List<HostAndPort>> getPartsLocation(String spaceName) {
+        Map<Integer, List<HostAndPort>> result;
+        try {
+            result = getPartsAlloc(spaceName);
+        } catch (ExecuteFailedException | TException e) {
+            return Maps.newHashMap();
+        }
+        return result;
+    }
+
+    /**
+     * get all tags info
+     *
+     * @param spaceName nebula graph space
+     * @return empty list if exception happen
+     */
+    private List<TagItem> listTags(String spaceName) {
+        List<TagItem> tagItems;
+        try {
+            tagItems = getTags(spaceName);
+        } catch (TException | ExecuteFailedException e) {
+            return Lists.newLinkedList();
+        }
+        return tagItems;
+    }
+
+    /**
+     * get all edges info
+     *
+     * @param spaceName nebula graph space
+     * @return empty list if exception happen
+     */
+    private List<EdgeItem> listEdges(String spaceName) {
+        List<EdgeItem> edgeItems;
+        try {
+            edgeItems = getEdges(spaceName);
+        } catch (TException | ExecuteFailedException e) {
+            return Lists.newLinkedList();
+        }
+        return edgeItems;
+    }
+
+    /**
+     * check if space exist
+     */
+    private boolean existSpace(String spaceName) {
+        List<IdName> spaces = listSpaces();
+        for (IdName space : spaces) {
+            if (new String(space.getName()).equals(spaceName)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * check if tag exist
+     */
+    private boolean existTag(String spaceName, String tag) {
+        if (!existSpace(spaceName)) {
+            return false;
+        }
+        List<TagItem> tags = null;
+        try {
+            tags = getTags(spaceName);
+        } catch (Exception e) {
+            LOGGER.error("failed to get tags", e);
+            return false;
+        }
+        for (TagItem item : tags) {
+            if (new String(item.getTag_name()).equals(tag)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * check if edge exist
+     */
+    private boolean existEdge(String spaceName, String edge) {
```

Comment:
diito

Reply:
the method is called only when cache doesn't contain edge.